softwareComponents/stm32cxx/src/drivers/i2c.hpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <array>
+#include <cstdint>
+#include <functional>
+
+#include <system/assert.hpp>
+
+#include <drivers/peripheral.hpp>
+#include <drivers/gpio.hpp>
+
+#include <stm32g0xx_ll_i2c.h>
+
+
+// TODO: rework Pin types to be avoid being interchangable
+struct SdaPin : public Gpio::Pin
+{
+    SdaPin( Gpio::Pin pin )
+        : Gpio::Pin( pin ) {}
+};
+
+struct SclPin : public Gpio::Pin
+{
+    SclPin( Gpio::Pin pin )
+        : Gpio::Pin( pin ) {}
+};
+
+struct I2C: public Peripheral< I2C_TypeDef > {
+    I2C( I2C_TypeDef *i2c, SdaPin sdaPin, SclPin sclPin )
+        : Peripheral< I2C_TypeDef >( i2c )
+    {
+        sdaPin.setupODAlternate( true );
+        sclPin.setupODAlternate( true );
+
+        // enableClock();
+        LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_I2C2);
+
+        LL_I2C_InitTypeDef I2C_InitStruct = { };
+
+        I2C_InitStruct.PeripheralMode = LL_I2C_MODE_I2C;
+        // This parameter is computed by STM32cubemx tool
+        I2C_InitStruct.Timing = 0x00303D5B;
+        I2C_InitStruct.AnalogFilter = LL_I2C_ANALOGFILTER_ENABLE;
+        I2C_InitStruct.DigitalFilter = 0;
+        I2C_InitStruct.OwnAddress1 = 0;
+        I2C_InitStruct.TypeAcknowledge = LL_I2C_ACK;
+        I2C_InitStruct.OwnAddrSize = LL_I2C_OWNADDRESS1_7BIT;
+
+        LL_I2C_Init( _periph, &I2C_InitStruct );
+        LL_I2C_EnableAutoEndMode( _periph );
+        LL_I2C_SetOwnAddress2( _periph, 0, LL_I2C_OWNADDRESS2_NOMASK );
+        LL_I2C_DisableOwnAddress2( _periph );
+        LL_I2C_DisableGeneralCall( _periph );
+        LL_I2C_EnableClockStretching( _periph );
```

Comment:
Toto bude potřeba výhledově zobecnit, aby to pracovalo s libovolným I2C zařízením. Ale hádám, že toho jsi si vědom a budeme to dál rozvíjet.